rfc/0003-policy-signing.md

```diff
@@ -0,0 +1,440 @@
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+> Why should we **not** do this?
+>
+>   * obscure corner cases
+>   * will it impact performance?
+>   * what other parts of the product will be affected?
+>   * will the solution be hard to maintain in the future?
+
+Policy signature verification is going to be an opt-in feature.
+
+There's no reason to not implement this feature.
+
+# Alternatives
+[alternatives]: #alternatives
+
+> - What other designs/options have been considered?
+> - What is the impact of not doing this?
+
+## Specify per-policy signature settings
+
+We could ask operators to put the signature verification settings
+inside of the `ClusterAdmissionPolicy` definition.
+
+For example:
+
+```yaml
+apiVersion: policies.kubewarden.io/v1alpha2
+kind: ClusterAdmissionPolicy
+metadata:
+  name: privileged-pods
+  annotations:
+    subject.cosign.sigstore.dev: https://github.com/kubewarden/privileged-pods/.github/workflows/release.yml@refs/tags/v0.1.9
+    issuer.cosign.sigstore.dev: https://token.actions.githubusercontent.com
+spec:
+  module: registry://ghcr.io/kubewarden/policies/pod-privileged:v0.1.9
+  rules:
+  - apiGroups: [""]
+    apiVersions: ["v1"]
+    resources: ["pods"]
+    operations:
+    - CREATE
+    - UPDATE
+  mutating: false
+```
+
+The same could be done to embed the public key to be used.
+
+We do not feel comfortable with this behaviour:
+
+  * The security settings can be provided by whoever has the RBAC rights to create/modify `ClusterAdmissionPolicy`
+    objects.
+  * The UX is still ugly for policies that signed inside of GitHub Actions using the non-interactive mode
+  * This approach can be applied only by Policy Server, kwctl would have to verify policies in a different way because it
+    doesn't rely on `ClusterAdmissionPolicy` objects to run/pull policies.
+
+## Embed the verification criteria inside of policy's metadata
+
+Each Kubewarden policy has some user-defined metadata.
+
+The process to publish a policy consists of the following steps:
+
+  1. Edit the `metadata.yml` file
+  2. Annotate the policy with the `metadata.yml` file. This produces a new `.wasm` file, which includes the metadata
+  3. Push the policy to a registry
+  4. Sign the policy using the immutable reference obtained inside of the previous step
+
+We could add a new entry inside of policy's metadata, something like:
+
+```yaml
+subject.cosign.sigstore.dev: https://github.com/kubewarden/privileged-pods/.github/workflows/release.yml@refs/tags/v0.1.9
+issuer.cosign.sigstore.dev: https://token.actions.githubusercontent.com
+```
+
+At verification time, we could look for a signature that satisfies the constrains provided inside of the policy metadata.
+
+Alterning the policy's metadata would lead to a different `.wasm` file, which would
+break the initial checks of sigstore's signature.
+
+We've discarded this approach because, among other things, we think it's safer to provide the verification constraints
+in an out-of-band way.
```

Comment:
You could add that it makes having reproducible builds of policy images more difficult, if not impossible.

Reply:
added